### ISSUE TITLE
Feature explanation: symmetries exploitation

### ISSUE THREAD

**Issue by Vandertic** [CuriosAI/sai#6]
I believe symmetries have been added also to the current next branch of LZ, but i did not check if the implementation is the same as our own.
What we have done in SAI is [here](https://github.com/sai-dev/sai/commit/86b8324e6bef5b493b5c16dac1547e1809191e39).

The new line-command option --symm modifies the agent behaviour in the following way.

For every node in the search tree the policy probabilities of moves which are equivalent by symmetry are
summed up and concentrated onto a single one of them, which is randomly chosen. The other symmetrical moves are completely excluded from the UC tree.

The same option modifies the training info in the following way.

The total number of visits for the chosen move, is split again (evenly, this time) among symmetry-equivalent moves at the moment of writing the training information.

In our experiments on 7x7 we observed that the progression at the beginning was much faster with symmetries enabled and that the optimal first move (tengen) is typically discovered sooner and with a lower number of visits than before.

Notice that this will affect the probabilities appearing in the training info, since instead of being the ratio v/N between move visits and total visits, for symmetric moves it may be V/N/2^n, where V is the total number of moves for the symmetry equivalent class and n is 1, 2 or 3 according to the order of the symmetry involved. So one cannot simply deduce the number of visits from the common denominator of the probabilities any more.

**Comment by alreadydone** [comment]
LZ simply looks for symetrical positions in NNCache and use it to expand a symmetrical node; it doesn't merge symmetrical nodes into one as you do. The LZ way should save more computation but yours save more memory; yours is akin to a transposition table.

**Comment by Nazgand** [comment]
Because SAI is not trained with a single non-zero komi, color-switching symmetries could also be exploited. This may let the neural network learn that passing is usually not a good idea more quickly, for most color-switching symmetric positions in the same game would be the result of a pass.

Checking a ko-move violation of symmetry before everything else would probably be slightly more efficient.

**Comment by Vandertic** [comment]
> @alreadydone: LZ simply looks for symetrical positions in NNCache and use it to expand a symmetrical node; it doesn't merge symmetrical nodes into one as you do. The LZ way should save more computation but yours save more memory; yours is akin to a transposition table.

Another difference is that if the visit number is fixed (say n), then LZ solution computes the network for a fraction m of the n nodes it visits and explores a wider tree. SAI solution computes the network for n nodes and explores a deeper tree. The exploration of LZ is much faster, at the price that if one identifies equivalent branches of the explored tree, on gets a smaller efficient tree with only m nodes.

**Comment by Vandertic** [comment]
> @Nazgand: Because SAI is not trained with a single non-zero komi, color-switching symmetries could also be exploited. This may let the neural network learn that passing is usually not a good idea more quickly, for most color-switching symmetric positions in the same game would be the result of a pass.
> 
> Checking a ko-move violation of symmetry before everything else would probably be slightly more efficient.

In fact, SAI network doesn't even use the color of the current player as an input, as explained [here](https://github.com/sai-dev/sai/issues/5), so color symmetry is hard coded in the net.

**Comment by alreadydone** [comment]
> Another difference is that if the visit number is fixed (say n), then LZ solution computes the network for a fraction m of the n nodes it visits and explores a wider tree. SAI solution computes the network for n nodes and explores a deeper tree. The exploration of LZ is much faster, at the price that if one identifies equivalent branches of the explored tree, on gets a smaller efficient tree with only m nodes.

True; LZ's symmetry exploitation (probe_cache) is disabled during self-play (fixed visits) though, so it neither explores deeper nor faster.

**Comment by Nazgand** [comment]
@Vandertic I read that, and understand that SAI does not know which stones are white and which stones are black, only which stones are the current player's, and which are the opponent's.  Identifying color-swap symmetries in the game tree and deduplicating them will save space, similar to how the rotation and mirror symmetry deduplication saves space.

The game PASS A4 is symmetric to the game A4 through color-swapping. alpha and beta should be the same because SAI does not know komi; only the color of the current player changed.

A way to speed up the symmetry detection is to have FastBoard or FullBoard variables for the number of current player stones and the number of opponent stones on the board. Knowing that the numbers of stones does not match would mean not needing to check half the vertices of the board before finding a non-matching vertex.

E.g. Detecting 180 rotation color-swap symmetry may give SAI a higher probability of learning that passing is not a good idea in most mirror Go situations.
